### DISCUSSION TITLE
[SQLite] - Conflict statements returning odd outputs with `inserted_primary_key`

### DISCUSSION THREAD

**Discussion by pm5k** [sqlalchemy/sqlalchemy#10512] [Usage Questions] [answered]
Hi,

I am currently facing some weird-looking behaviour when working with SQLAlchemy and SQLite.
I have set up a context manager for my sessions which looks like this (it's fairly standard tbh..):

```py
@contextmanager
def session() -> Generator[Session, None, None]:
    session = SessionLocal()
    try:
        yield session
        session.commit()
    except SQLAlchemyError:
        session.rollback()
        raise
    finally:
        session.close()
```

And then I decided to write an upsert function which deals with three objects in unison. However the behaviour is present in all three parts of the function, so for brevity - I am only going to show one part. None of those three are coupled, so using it in isolation also produces this result.

### Context:
The first three items in the list below already exist in the db and would cause a unique constraint conflict on "name".
The last one should come through just fine.

### Upsert function:
```py
# this method should be called inside a session context
# which then passes the session object as an arg.

def myupsert(session: Session):
  # ... 
  for tag in ["foo", "bar", "baz", "foobar"]:
        tag_stmt = (
            insert(Tag)
            .values({"name": tag})
            .on_conflict_do_nothing(index_elements=["name"])
        )
        result = session.execute(tag_stmt)
        print(">>>> ", result.inserted_primary_key)
  # ...
```

### Output:
```
(9,)
(9,)
(9,)
(4,) <-- actually a legit ID for the newly inserted non-conflicting tag
```

### Issue:
There are a few questions that spring to mind when seeing this output. For instance:

- Why am I seeing `9` as the returned primary key when a conflict occurs? It would make all the sense in the world to give a `None` or even some `ConflictErrorObj` to explicitly state this was a failure, although in my head the most desired outcome is to receive a tuple which tells you that there was a conflict and at which PK for instance - `(True, 1)`. Why the `9`?
- Upon further inspection, the results returning 9 are also counted as closed, so for instance if I try to execute a `fetchall()` on the result of that execute statement I will see `sqlalchemy.exc.ResourceClosedError: This result object does not return rows. It has been closed automatically`. Reading further, the docs explain a bit about this behaviour [here](https://docs.sqlalchemy.org/en/13/core/connections.html#dbengine-implicit), but I have tried a direct engine execute outside of the session scope and the results I saw were identical, something just closes that right after execution making me unable to retrieve anything useful from the object touched by that execute.

### Expected behaviour:
I honestly did not know what to expect, all I wanted to do was to see if I can avoid issuing post-execute queries to retrieve the ID's of the objects from that cursor result so I could then use them further down in that function, but getting 9's back and being unable to otherwise refer to the object necessitates additional query work. I would like someone with knowledge around this to explain to me if I am just doing something dumb/expecting something unreasonable here.

Additionally, it concerns me that that `9` is returned, since that could really mess up logic for folks who may want to grab conflict record id's in a loop and the code acts on them somehow. Imagine I filtered on `9` to avoid that and a legitimate record got added for ID 9 - now I would end up filtering that out in error. 

Anyway, I am keen to hear what you guys think of this and welcome any guidance.

**Comment by CaselIT** [comment, marked as answer]
Hi,

This seems just how the sqlite3 driver for python works. What's returned is the value of the autoincrementing sequence. Example without using sqlalachemy:
```py
import sqlite3

c = sqlite3.connect(':memory:')
c.execute('''
CREATE TABLE tbl (
        id INTEGER NOT NULL,
        name TEXT,
        PRIMARY KEY (id),
        UNIQUE (name)
)''')
c.executemany('INSERT INTO tbl (name) VALUES (?)', [('foo',) ,('bar',), ('baz',)])

cursor = c.execute('INSERT INTO tbl (name) VALUES (?) ON CONFLICT (name) DO NOTHING', ('foo',))
print(cursor.lastrowid)
cursor = c.execute('INSERT INTO tbl (name) VALUES (?) ON CONFLICT (name) DO NOTHING', ('new',))
print(cursor.lastrowid)
c.execute('DELETE FROM tbl WHERE name = ?', ('bar',))
cursor = c.execute('INSERT INTO tbl (name) VALUES (?) ON CONFLICT (name) DO NOTHING', ('foo',))
print(cursor.lastrowid)
```
this prints
```
3
4
4
```

> so for instance if I try to execute a fetchall() on the result of that execute statement I will see sqlalchemy.exc.ResourceClosedError: This result object does not return rows. It has been closed automatically. Reading further, the docs explain a bit about this behaviour [here](https://docs.sqlalchemy.org/en/13/core/connections.html#dbengine-implicit), but I have tried a direct engine execute outside of the session scope and the results I saw were identical, something just closes that right after execution making me unable to retrieve anything useful from the object touched by that execute.

The statement you are running does not return anything, so I'm not sure what you would expect to get when calling `fetchall()`.

> although in my head the most desired outcome is to receive a tuple which tells you that there was a conflict and at which PK for instance - (True, 1)

use returning for that:
```py
tag_stmt = insert(Tag).values({"name": tag}).on_conflict_do_nothing(index_elements=["name"]).returning(Tag.id)
result = session.execute(tag_stmt)
print(">>>> ", result.first())
```

returning does not return any row in case of conflict, so `first()` returns `None`

**Comment by zzzeek** [comment]
it doesnt really make sense to use `inserted_primary_key` for an upsert since these rows were not inserted.    The attribute makes use of the DBAPI's `.lastrowid` element here which would have undefined behavior in this case.    as mentioned, always use RETURNING to get an accurate view of the affected row.

**Comment by pm5k** [comment]
@CaselIT and @zzzeek thank you both, this works.
